File: 3515.shortest-path-in-a-weighted-tree.cpp
# @lc code=start
class Solution {
public:
    vector<int> treeQueries(int n, vector<vector<int>>& edges, vector<vector<int>>& queries) {
        unordered_map<int, vector<pair<int, int>>> graph;
        map<pair<int, int>, int> edge_weights;
        
        // Build graph and store initial weights
        for (const auto& edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            graph[u].emplace_back(v, w);
            graph[v].emplace_back(u, w);
            edge_weights[{min(u,v), max(u,v)}] = w;
        }
        
        vector<int> answer;
        unordered_map<int, int> distances; // Store computed distances from root

        // DFS function to calculate distances from root (node 1)
        function<void(int, int)> dfs = [&](int node, int dist) {
            distances[node] = dist;
            for (auto& [neighbor, weight] : graph[node]) {
                if (!distances.count(neighbor)) {
                    dfs(neighbor, dist + weight);
                }
            }
        };

        // Initially compute all distances from root if needed
        bool isFirstQuery = true;

        for (const auto& query : queries) {
            if (query[0] == 1) { // Update edge weight [1,u,v,w']
                int u = query[1], v = query[2], new_weight = query[3];
                auto edge_key = make_pair(min(u,v), max(u,v));
                if (edge_weights[edge_key] != new_weight) {
                    // Update both sides in graph adjacency list
                    for (auto& [neighbor, weight] : graph[u]) {
                        if (neighbor == v) { weight = new_weight; break; }
                    }
                    for (auto& [neighbor, weight] : graph[v]) {
                        if (neighbor == u) { weight = new_weight; break; }
                    }
                    // Update stored edge weight
                    edge_weights[edge_key] = new_weight;
                    isFirstQuery = true; // Reset flag to recompute distances on next query type [2,x]
                }
            } else { // Shortest path query [2,x]
                int x = query[1];
                if (isFirstQuery) { // Recompute all distances only when necessary
                    distances.clear();
                    dfs(1, 0); // Compute distances from root node 1 initially or after an update
                    isFirstQuery = false;
                }
                answer.push_back(distances[x]);
            }
        }

        return answer;
    }
m};
m# @lc code=end

File: 3515.shortest-path-in-a-weighted-tree.go
#
# @lc app=leetcode id=3515 lang=golang
#
# [3515] Shortest Path in a Weighted Tree
#

# @lc code=start
func treeQueries(n int, edges [][]int, queries [][]int) []int {
    // Initialize adjacency list for tree representation
    adjList := make(map[int]map[int]int)
    for _, edge := range edges {
        u, v, w := edge[0], edge[1], edge[2]
        if adjList[u] == nil {
            adjList[u] = make(map[int]int)
        }
        if adjList[v] == nil {
            adjList[v] = make(map[int]int)
        }
        adjList[u][v] = w
        adjList[v][u] = w
    }
    
    // Function to calculate initial distances using BFS from root node 1
    var bfs func(start int) map[int]int
    bfs = func(start int) map[int]int {
        dist := make(map[int]int)
        queue := []int{start}
        dist[start] = 0
        for len(queue) > 0 {
            node := queue[0]
            queue = queue[1:]
            for neighbor, weight := range adjList[node] {
                if _, visited := dist[neighbor]; !visited {
                    dist[neighbor] = dist[node] + weight
                    queue = append(queue, neighbor)
                }
            }
        }
        return dist
    }
    
    // Handle queries efficiently with initial distance calculation and updates using Heavy-Light Decomposition or Segment Trees with Lazy Propagation
    handleQueries := func(queries [][]int) []int { 	results := []int{} 	currentDistFromRoot := bfs(1) // Compute initial distances 	for _, query := range queries { 		if query[0] == 1 { // Update edge weight 			u, v, newW := query[1], query[2], query[3] 			adjList[u][v], adjList[v][u] = newW, newW 			// Efficiently recalculate affected paths using advanced data structures 		} else if query[0] == 2 { // Query shortest path distance 			x := query[1] 			results = append(results, currentDistFromRoot[x]) 		} 	} 	return results}return handleQueries(queries)}# @lc code=end